Generate shim in build

wli3 opened the pull request:
Continue of https://github.com/dotnet/sdk/pull/2366

problem i know so far

* pending the solution to use cached asset file

It supports Clean target. However, it will not remove empty folder. Since different RID's shims have the same name. I need to separate them by folder. But msbuild will not remove that https://github.com/Microsoft/msbuild/issues/2408

Commit:
Generate shim in build

When shim RID is set

Commit:
dotnet clean can remove shims

wli3 commented:
> Incremental performance: we did a lot of work to avoid reading the assets file from json on every incremental build, we need to preserve that here. Ideally, we would not even read the assets file or regenerate a shim if the things that affect the content of shim have not changed.

I tried to directly use `NativeLibraries` itemgroup and I realize it includes 1 RID at most. And generate shim need multiple RID's apphost. I think the solution could be

a. Resolve and add these assets during ResolvePackageAssets. And write them into cache.(or even just output an extra itemgroup). It should be best for perf, but it will introduce implicit coupling of the 2 tasks.

b. Add logic to detect "should rewrite shim". It needs to detect the input params change. It needs another file to record what we write into shim last time (we cannot read back from shim).

We could do a and b both

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateShims.cs:
```diff
@@ -81,6 +81,7 @@ public sealed class GenerateShims : TaskBase
         [Output]
         public ITaskItem[] EmbeddedApphostPaths { get; private set; }
 
+
```
nit: newline added

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets:
```diff
@@ -68,13 +75,26 @@ Copyright (c) .NET Foundation. All rights reserved.
+    -->
+
+  <Target Name="GenerateShimsAssets"
+          AfterTargets="ResolveReferences"
```
`AfterTargets` is fine, with the caveat that you shouldn't do `AfterTargets="Build"` because of https://github.com/Microsoft/msbuild/issues/3345. But I generally prefer to use `BeforeTargets`--can you `BeforeTargets="IncrementalClean"`?

wli3 replied:
@rainersigwald thank, although the task itself depends on `ResolveReferences`. Should I just dependson =`ResolveReferences`. and `BeforeTargets="IncrementalClean"`. But I think use dependson with before/after target is kind of strange.

rainersigwald replied:
It's reasonable to have both `BeforeTargets="IncrementalClean" DependsOnTargets="ResolveReferences"`. I'd go with that.

It is funny looking, but they're different relationships: BeforeTargets means "before you run that, run me first", while DependsOn means "before I run, run these first". Both can be true at once.

nguerrera replied:
I would expect the logic for this target to be the same as GenerateBuildDependencyFile. We went through a few iterations on that and ended up against beforetargets=incrementalclean IIRC. It would be good to have one pattern for extra files we create.

wli3 replied:
I see, I read the comment. It makes sense. Now the problem is, how to set the timing(before or after what?) I think `BeforeTargets="CopyFilesToOutputDirectory"` is reasonable.

Changed to following, I will push later but I don't want to hide the thread.

```xml
  <Target Name="GenerateShimsAssets"
          BeforeTargets="CopyFilesToOutputDirectory"
          DependsOnTargets="ResolveReferences"
          Condition="'$(PackAsTool)' == 'true'" >

    <ItemGroup>
      <_PackAsToolShimRuntimeIdentifiers Include="$(PackAsToolShimRuntimeIdentifiers)"/>
    </ItemGroup>

    <PropertyGroup>
      <PackagedShimOutputRootDirectory Condition=" '$(PackagedShimOutputRootDirectory)' == '' ">$(OutDir)</PackagedShimOutputRootDirectory>
    </PropertyGroup>

    <GenerateShims
      DotNetAppHostExecutableNameWithoutExtension="$(_DotNetAppHostExecutableNameWithoutExtension)"
      PackagedShimOutputDirectory="$(PackagedShimOutputRootDirectory)/shims/$(TargetFramework)"
      PackageId="$(PackageId)"
      PackageVersion="$(Version)"
      ProjectAssetsFile="$(ProjectAssetsFile)"
      ProjectPath="$(MSBuildProjectFullPath)"
      ShimRuntimeIdentifiers="@(_PackAsToolShimRuntimeIdentifiers)"
      TargetFrameworkMoniker="$(NuGetTargetMoniker)"
      ToolCommandName="$(ToolCommandName)"
      ToolEntryPoint="$(ToolEntryPoint)"
      >

      <Output TaskParameter="EmbeddedApphostPaths" ItemName="_EmbeddedApphostPaths" />
    </GenerateShims>

    <ItemGroup>
      <!-- Do this in an ItemGroup instead of as an output parameter so that it still gets added to the item set
           during incremental builds when the task is skipped -->
      <FileWrites Include="@(_EmbeddedApphostPaths)"/>
    </ItemGroup>
  </Target>
```

Commit:
Remove AfterTargets

Commit:
Do not use output parameter for clean

wli3 commented:
> > Incremental performance: we did a lot of work to avoid reading the assets file from json on every incremental build, we need to preserve that here. Ideally, we would not even read the assets file or regenerate a shim if the things that affect the content of shim have not changed.
> 
> I tried to directly use `NativeLibraries` itemgroup and I realize it includes 1 RID at most. And Generateshim need multiple RID's apphost. I think the solution could be
> 
> a. Resolve and add these assets during ResolvePackageAssets. And write them into cache.(or even just output an extra itemgroup). It should be best for perf, but it will introduce implicit coupling of the 2 tasks.
> 
> b. Add logic to detect "should rewrite shim". The complex part is, we need to detect the input params change. We needs another file to record what we write into shim last time (due to we cannot read back from shim).
> 
> We could do a and b both

@nguerrera what do you think of the asset file problem

Commit:
Use asset file cache and enable incremental build

wli3 commented:
@nguerrera 
1. I added these apphost assets to ResolvePackageAssets and cached it.
2. thanks to @rainersigwald  I used method similar to [this](https://github.com/Microsoft/msbuild/blob/7d8c8da147252ea57f35d2c6ceee8e7c5ba458f5/src/Tasks/Microsoft.Common.CurrentVersion.targets#L3408-L3435), to trigger incremental build when property change.

Commit:
Add workaround for setLastWriteTime precision

Commit:
Add params to hash

wli3 commented:
@nguerrera ping review

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ExecutableExtension.cs:
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
```
nit: extra newline

wli3 replied:
fixed

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ExecutableExtension.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public sealed class ExecutableExtension
+    {
+        public static string GetExecutableExtensionAccordingToRuntimeIdentifier(string runtimeIdentifier)
```
I wonder if we can avoid this altogether by just preserving the apphost extension (or lack thereof). In other words, if we find apphost.exe, we make mytool.exe. If we find apphost, we make mytool. (And more theoretically, if we find apphost.future_os_extension, we make mytool.future_os_extension,)

wli3 replied:
Plus the comment below. It has problem with GetEmbeddedApphostPaths(a computer only, get where to look for these files). In this we don't require template apphost path. I think it is over kill to depend on template apphost path just to get the extension.

I renamed it to ForRuntimeIdentifier

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PublishRelativePathToPackPackagePath.cs:
```diff
@@ -0,0 +1,110 @@
+        public ITaskItem[] ResolvedFileToPublishWithPackagePath { get; private set; }
+
+        private const char DirectorySeparatorChar = '\\';
+        private const char AltDirectorySeparatorChar = '/';
```
These are confusing me since they seem like they might just be `using static Path` when reading the code. These names are generally OS-specific. How about rename to ForwardSeparatorChar and BackwardSeparatorChar. Or WindowsSeparatorChar and UnixSeparatorChar?

wli3 replied:
fixed, renamed to ForwardSeparatorChar and BackwardSeparatorChar

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -866,6 +894,32 @@ private void WriteNativeLibraries()
+                    if (runtimeIdentifier.StartsWith("win"))
+                    {
+                        apphostName += ".exe";
+                    }
```
I see now that we search for the full name, not the name without extension and the approach I thought about earlier where we just preserve the extension could be prone to breaking if there's ever an apphost.something_else in the package.

wli3 replied:
fixed, changed to use helper.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets:
```diff
@@ -68,30 +67,106 @@ Copyright (c) .NET Foundation. All rights reserved.
+      <!-- Do this in an ItemGroup instead of as an output parameter so that it still gets added to the item set
+           during incremental builds when the task is skipped -->
+      <FileWrites Include="@(_EmbeddedApphostPaths)" Condition="Exists('@(_EmbeddedApphostPaths)')"/>
+      <FileWrites Include="$(_ShimCreatedSentinelFile)" Condition="Exists('$(_ShimCreatedSentinelFile)')"/>
```
I didn't know that Exists works on items. What does it return if one exists, but another doesn't.

I don't think it's necessary to condition adds to filewrites

wli3 replied:
removed condition

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets:
```diff
@@ -68,30 +67,106 @@ Copyright (c) .NET Foundation. All rights reserved.
+         LastWriteTime in shims are not accurate. And _ShimInputCacheFile will have later timestamp than generated shims.
+         Use a created file to "record" LastWriteTime. And only use it in "Outputs" field for timestamp comparison.
+    -->
+    <WriteLinesToFile Lines="This file's LastWriteTime is used in incremental build" File="$(_ShimCreatedSentinelFile)" Overwrite="True" />
```
Consider using Touch task for this, which is more idiomatic in msbuild for this purpose. You'd lose this comment in the file, but I don't think we need it.

wli3 replied:
Touch uses the same corefx api that is having the bug. Once it is fixed, we can remove all this.
https://github.com/Microsoft/msbuild/blob/4972ebc6d2b81c5e967cd3260943a95c8e5eb4be/src/Tasks/Touch.cs#L133

nguerrera replied:
Can you just add a comment that Touch uses the same API and therefore has the same problem.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PublishRelativePathToPackPackagePath.cs:
```diff
@@ -0,0 +1,110 @@
+    /// 1. Publish PublishRelativePath is a relative <strong>file</strong> path. This need to convert to relative path.
+    /// 2. Due to DotnetTools package format, "tools/{TargetFramework}/any/" is needed in addition to the relative path.
+    /// </summary>
+    public sealed class PublishRelativePathToPackPackagePath : TaskBase
```
This name sounds like it's going to publish something and also doesn't make it very clear that it's specific to getting the paths for a DotnetTools package

Commit:
Address code review

- Add comment for setting SetLastWriteTimeUtc

- Rename and ignore rid casing

- Use ExecutableExtension.ForRuntimeIdentifier

- Rename and remove condition on include

- Rename to RelativePathFormatInPublishToFormatInPackPackage

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs:
```diff
@@ -61,6 +56,8 @@ public static class AppHost
                 }
             }
+
+            File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
```
Added comment

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/ExecutableExtension.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        public static string GetExecutableExtensionAccordingToRuntimeIdentifier(string runtimeIdentifier)
+        {
+            if (runtimeIdentifier.StartsWith("win"))
```
fixed

[insensitive](https://github.com/dotnet/cli/pull/8997#pullrequestreview-109813807 ). I added OrdinalIgnoreCase

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets:
```diff
@@ -68,30 +67,106 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <WriteLinesToFile Lines="$(_GenerateShimsAssetsInputCacheHash)" File="$(_ShimInputCacheFile)" Overwrite="True" WriteOnlyWhenDifferent="True" />
+
+    <ItemGroup>
+      <FileWrites Include="$(_ShimInputCacheFile)" Condition="Exists('$(_ShimInputCacheFile)')" />
```
Removed

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets:
```diff
@@ -24,34 +29,28 @@ Copyright (c) .NET Foundation. All rights reserved.
       <_GeneratedFiles Include="$(_ToolsSettingsFilePath)"/>
     </ItemGroup>
 
+    <RelativePathFormatInPublishToFormatInPackPackage
```
This name is still confusing me. Can it be something like `ResolveToolPackagePaths`?

Commit:
Rename to ResolveToolPackagePaths

Commit:
Added comment for why touch task cannot be used

wli3 commented:
@dotnet-bot test OSX10.12 Release

wli3 commented:
@dotnet-bot test OSX10.12 Release
